[lldb/crashlog] Implement speculative binary lookup for target creation

medismailben opened the pull request:
This patch changes the way the main executable gets looked up when creating the target in the crashlog script.

On macOS, when a process crashes while being built from a user home directory, the path to the binary gets obfuscated in the crashlog to avoid revealing any user personal information like the username of even the user file system.

This safety feature breaks the main target creating lookup for crashlogs since the obfuscated path in the crashlog doesn't exists on disk.

This patch addresses this issue by speculatively patching the obfuscated binary patch and scanning every executable and subdirectories (in parallel) in said path, and looking for a matching UUID as the one provided in the crash report.

This usually enough to find the executable locally.

medismailben commented on lldb/examples/python/crashlog.py:
```diff
@@ -340,11 +341,101 @@ def find_matching_slice(self):
+                    for d in os.listdir(search_path)
+                    if os.path.isdir(os.path.join(search_path, d))
+                ]
+            except PermissionError:
```
Good point

bulbazord reviewed (approved):
I wouldn't say I'm an expert but I don't see any glaring issues here. LGTM

JDevlieghere reviewed (changes requested):
I don't think it's appropriate for the crashlog script to recurse through someone's home directory. This is going to result in TCC (Transparency Consent and Control) popups when trying to access certain directories (like Photos, Downloads, etc). That's by design and a good indication that this isn't a good idea. As a user, if I saw a script/program do this without my explicit direction to do so, I would be very unhappy.  

Potential alternatives that I think are fine:

- Replacing `/Users/USER` by `~` but presumably everything after is stripped to, which is why you're doing this in the first place. 
- Using mdfind to find the dSYM for the binary, and looking next to it, similar to what LLDB does. 
- Relying on `dsymForUUID` to give you the symbol rich executable.

bulbazord reviewed (changes requested):
Jonas makes a great point, I didn't consider the privacy aspect of this change.

I agree with him, I don't think this change should go in as-is.

medismailben committed:
[lldb/crashlog] Implement speculative binary lookup for target creation

This patch changes the way the main executable gets looked up when
creating the target in the crashlog script.

On macOS, when a process crashes while being built from a user home
directory, the path to the binary gets obfuscated in the crashlog to
avoid revealing any user personal information like the username of even
the user file system.

This safety feature breaks the main target creating lookup for crashlogs
since the obfuscated path in the crashlog doesn't exists on disk.

This patch addresses this issue by speculatively patching the obfuscated
binary patch and scanning every executable and subdirectories (in
parallel) in said path, and looking for a matching UUID as the one
provided in the crash report.

This usually enough to find the executable locally.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>
